src/js/structs/Batch.js

```diff
@@ -44,35 +44,44 @@ import TransactionTypes from '../constants/TransactionTypes';
 
 class Batch {
   constructor() {
-    const batch = [];
-
-    this.add = this.add.bind(batch);
-    this.reduce = this.reduce.bind(batch);
+    concealContext(this, []);
   }
 
   /**
    * Add an action to the batch.
    *
+   * @this Array
    * @param {Object} item
+   * @returns {Batch} Returns the new batch
    */
   add(item) {
+    // Operate on a new array
+    const batch = this.slice();
+
     // Remove previous if path is the same as current to minimize
     // number of actions
-    let {path} = this[this.length - 1] || {};
+    let {path} = batch[batch.length - 1] || {};
     let hasEqualPaths = item && Array.isArray(path) &&
       Array.isArray(item.path) && path.join() === item.path.join();
     if (item.type === TransactionTypes.SET && hasEqualPaths) {
-      this.pop();
+      batch.pop();
     }
 
-    this.push(item);
+    batch.push(item);
+
+    // Create a new batch, without calling the constructor
+    let newInst = Object.create(Batch.prototype);
```

Comment:
Yes, because if I use the constructor, the `add` and `reduce` will be already bound on an empty array :)